py/server/deephaven/table.py

```diff
@@ -62,6 +64,106 @@ def _j_py_script_session() -> _JPythonScriptSession:
         return None
 
 
+def _analyze_func_signature(parameter_types: List[np.dtype] = None, return_type: np.dtype = None) -> str:
+    if all([parameter_types, return_type]):
+        np_type_codes = [np.dtype(p).char for p in parameter_types]
+        np_type_codes.extend(["-", ">", np.dtype(return_type).char])
+        return "".join(np_type_codes)
+    elif not any([parameter_types, return_type]):
+        ...
+    elif return_type is None:
+        raise ValueError("Missing target function return type.")
+    else:
+        raise ValueError("Missing target function parameter type(s)")
+
+
+def _chunked(parameter_types: List[np.dtype] = None, return_type: np.dtype = None):
+    """A decorator to wrap a Python function used in Deephaven query formulas and invoked on a row basis. The wrapper
+    function is able to accept arrays of the declared parameter data type (vs. scalar), thus minimize the Java/Python
+    boundary crossings.
+
+    Args:
+        parameter_types: ...
+        return_type: ...
+    """
+    np_type_codes = _analyze_func_signature(parameter_types=parameter_types, return_type=return_type)
+
+    def decorator(fn: Callable):
+        @wraps(fn)
+        def wrapper(*args):
+            if len(args) != len(parameter_types):
+                raise ValueError("The number of arguments doesn't match the definition.")
+
+            chunk_size = len(args[0])
+            chunk_result = np.empty(chunk_size, return_type)
+            i = 0
+            for scalar_args in zip(*args):
+                chunk_result[i] = fn(*scalar_args)
+                i += 1
+            return chunk_result
+
+        wrapper.dh_chunked = True
+        wrapper.signature = np_type_codes
+        return wrapper
+
+    return decorator
+
+
+def encode_signature(fn: Callable) -> str:
+    try:
+        sig = inspect.signature(fn)
+    except:
+        return ""
```

Comment:
`inspect.signature()` seems to differentiate between `ValueError` for no signature can be provided and `TypeError` for the type of the object is not supported.  Are we losing important information by treating them both as `""`?

Reply:
I don't think so. The type hint of fn clearly calls for an Callable, also it is meant for internal use, I think it is safe to ignore the difference between ValueError and TypeError here